File: DDD DSL/DDD DSL Design War Measurement System/Code/Java/src/fayelab/ddd/measure/arithmetic/MeasureSystem.java
package fayelab.ddd.measure.arithmetic;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.IntStream;

public class MeasureSystem
{
    private List<Integer> stepFactors;
    private List<Integer> baseFactors;

    public MeasureSystem(List<Integer> stepFactors)
    {
        this.stepFactors = stepFactors;
        this.baseFactors = makeBaseFactors(stepFactors);
    }

    public int base(List<Integer> quantityVec)
    {
        return VecOp.dotProduct(quantityVec, baseFactors);
    }

    public boolean equal(List<Integer> quantityVec1, List<Integer> quantityVec2)
    {
        return base(quantityVec1) == base(quantityVec2);
    }

    public List<Integer> normalize(int quantityValue)
    {
        return VecOp.dotDiv(quantityValue, baseFactors);
    }

    public List<Integer> add(List<Integer> quantityVec1, List<Integer> quantityVec2)
    {
        return normalize(base(quantityVec1) + base(quantityVec2));
    }
    
    public List<Integer> scale(int c, List<Integer> quantityVec)
    {
        return normalize(c * base(quantityVec));
    }

    private List<Integer> makeBaseFactors(List<Integer> stepFactors)
    {
        List<Integer> result = new ArrayList<>();
        IntStream.range(1, stepFactors.size())
                 .collect(() -> result,
                          (acc, idx) -> {
                              List<Integer> tail = stepFactors.subList(idx, stepFactors.size());
                              acc.add(tail.stream().reduce(1, (x, y) -> x * y));
                          },
                          (acc1, acc2) -> {
                              acc1.addAll(acc2);
                          });
        result.add(1);
        return result;
    }
    
    List<Integer> getStepFactors()
    {
        return stepFactors;
    }

    List<Integer> getBaseFactors()
    {
        return baseFactors;
    }
}

File: DDD DSL/DDD DSL Design War Measurement System/Code/Python/arithmetic/measuresystem.py
import vecop

def basefactors(stepfactors):
    result = []
    for (i, stepFactor) in enumerate(stepfactors[1:]):
        result.append(reduce(lambda x, y: x * y, stepfactors[i + 1:]))
    result.append(1)
    return result

def basefactors2(stepfactors):
    return reduce(_acc_basefactors2, stepfactors[::-1][:-1], [1])[::-1]

def _acc_basefactors2(acc, stepfactor):
    acc.append(acc[-1] * stepfactor)
    return acc

def base(quantityvec, basefactors):
    return vecop.dotproduct(quantityvec, basefactors)

def equal(quantityvec1, quantityvec2, basefactors):
    return base(quantityvec1, basefactors) ==  base(quantityvec2, basefactors)

def normalize(quantityvalue, basefactors):
    return vecop.dotdiv(quantityvalue, basefactors)

def add(quantityvec1, quantityvec2, basefactors):
    return normalize(base(quantityvec1, basefactors) + base(quantityvec2, basefactors), basefactors)

def scale(c, quantityvec, basefactors):
    return normalize(c * base(quantityvec, basefactors), basefactors)